org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/SourceTypeBinding.java

```diff
@@ -1634,33 +1634,19 @@ public long getAnnotationTagBits() {
 		return this.prototype.getAnnotationTagBits();
 
 	if ((this.tagBits & TagBits.AnnotationResolved) == 0 && this.scope != null) {
-		if ((this.tagBits & TagBits.EndHierarchyCheck) == 0) {
-			CompilationUnitScope pkgCUS = this.scope.compilationUnitScope();
-			boolean current = pkgCUS.connectingHierarchy;
-			pkgCUS.connectingHierarchy = true;
-			try {
-				initAnnotationTagBits();
-			} finally {
-				pkgCUS.connectingHierarchy = current;
-			}
-		} else {
-			initAnnotationTagBits();
+		TypeDeclaration typeDecl = this.scope.referenceContext;
+		boolean old = typeDecl.staticInitializerScope.insideTypeAnnotation;
+		try {
+			typeDecl.staticInitializerScope.insideTypeAnnotation = true;
+			ASTNode.resolveAnnotations(typeDecl.staticInitializerScope, typeDecl.annotations, this);
+		} finally {
+			typeDecl.staticInitializerScope.insideTypeAnnotation = old;
 		}
+		if ((this.tagBits & TagBits.AnnotationDeprecated) != 0)
+			this.modifiers |= ClassFileConstants.AccDeprecated;
```

Comment:
Can I have some commentary here ? Could `getAnnotationTagBits()` end up getting called too early to trigger issues ? What really is changing here ??

Reply:
Sure, this deserves a few words.

At face value this PR reverts several previous steps all going in a non-optimal direction, i.e., trying to call `getAnnotationTagBits()` wherever this might be needed and in turn let this method detect when it's being called "too early".

In particular, I'm reverting `getAnnotationTagBits()` to the state before https://git.eclipse.org/r/c/jdt/eclipse.jdt.core/+/147651/ (see also #1074).

The idea is, that all efforts now go into **avoiding** ahead-of-time calls, rather than trying to cope and invoke this code in a mode with deferred checking (see removed method `ClassScope.deferCheck()`.

If we are unsure, whether any call chains to `getAnnotationTagBits()` escaped my change of strategy, then we might add at least some kind of `assert` to detect this violation (as to aid us in hunting future bugs). Anyway, none of the tests from previous bug reports in this area show such a gap.